src/irr/asset/CGLSLBRDFBuiltinIncludeLoader.h

```diff
@@ -342,6 +344,25 @@ float irr_glsl_beckmann(in float a2, in float NdotH2)
     return irr_glsl_RECIPROCAL_PI * nom/denom;
 }
 
+//TODO this is taking into account beckmann NDF only, i'd like to also account for masking/shadowing term (visibility of microfacets)
+BSDFSample irr_glsl_beckmann_cos_gen_sample(in uvec2 _sample, in float _a2)
+{
+    vec2 u = vec2(_sample)/float(4294967295);
+    float phi = 2.0*irr_glsl_PI*u.x;
+    float tan2theta = u.y==1.0 ? 0.0 : (-_a2*log(1.0-u.y));
+    float cos2Theta = 1.0 / (1.0 + tan2theta);
+    float cosTheta = sqrt(cos2Theta);
+    float sinTheta = sqrt(1.0 - cos2Theta);
+    
+    BSDFSample smpl;
+    //erm... actually i'm returning half vector here, idk how could i get L with this api
+    //user should read this H (half vector/micro-surface normal) and compute L by reflecting V on H
+    //just remember that returned vector is in reflection space so first it has to be transformed to space in which light computations are done
+    smpl.L = vec3(sinTheta,sinTheta,cosTheta)*vec3(cos(phi),sin(phi),1.0);
```

Comment:
You need the frisvad tangent space creation function from example 22 (RaytracedAO, it might have been removed since then, but in one of the revisions it was there) this will allow you to transform out the half vector into whatever space `V` and `N` are

I suggest adding a struct `irr_glsl_AnisotropicViewSurfaceInteraction` to hold normalized `T`, `B`,`TdotV` and `BdotV`
Then this starts being relevant.
> hmm we could call `irr_glsl_NormalizedViewSurfaceInteraction` as `irr_glsl_IsotropicViewSurfaceInteraction` instead

We can have two overloads of `irr_glsl_calcAnisotropicInteraction(in irr_glsl_IsotropicViewSurfaceInteraction isotropic, args...)` :
- One where `args` are `in vec3 T, in vec3 B` (user knows and cares about tangent space, require T and B be normalized inputs)
- One where `args` is just `in vec3 T`, its the same as above but you construct bitangent from `cross(N,T)`
- One where `args` are nothing and you use Frisvad method to construct tangent space quickly

Then using T,B,N it will be really easy for you to bring the H vector to the space of `V`

Reply:
nice, thanks

Reply:
Actually `(TdotV,BdotV,NdotV)` is the transformation of `V` into the local tangent space where `N=(0,0,1)`

which means that reflection gets piss-easy
```glsl
// this should probably be implemented in our own HLSL-like `sincos`
float cosPhi = cos(phi);
float sinPhi = uintBitsToFloat(floatBitsToUint(sqrt(1.0-sinPhi*sinPhi))^(phi>irr_glsl_PI ? 0x80000000u:0u));

float TdotH = cosPhi*sinTheta;
float BdotH = sinPhi*sinTheta;
// putting these here for clarity, really you should be using these names instead
NdotH = cosTheta;

VdotH = TdotV*TdotH+BdotV*BdotH+NdotV*NdotH;
float twiceVdotH = 2.0*VdotH;

vec3 localH = vec3(TdotH,BdotH,NdotH);
L = (mat3(T,B,N)*localH)*twiceVdotH-V;

// freebies
VdotL = twiceVdotH*VdotH-1.0;
TdotL = twiceVdotH*TdotH-TdotV;
BdotL = twiceVdotH*BdotH-BdotV;
NdotL = twiceVdotH*NdotH-NdotV;
```

P.S. `BSDFSample` should probably return `TdotH`,`BdotH`,`NdotH`,`HdotV` and the freebies alongside the existing probability and `L`. In all honesty there should always be two overloads, one for isotropic output and one for anisotropic.

Reply:
Actually the part onwards from when you establish `TdotH`,`BdotH` and `NdotH` could be a helper function shared amongst all specular sampling routines since the rest of the math is always the same

Reply:
hm, i need this tangent frame (or whatever it's called, [T,B,N]) regardless of whether it's isotrpic or anisotropic distribution (i need to put L into same space where N is defined) so im not sure irr_glsl_AnisotropicViewSurfaceInteraction and irr_glsl_IsotropicViewSurfaceInteraction is the right solution

Reply:
> hm, i need this tangent frame (or whatever it's called, [T,B,N]) regardless of whether it's isotrpic or anisotropic distribution (i need to put L into same space where N is defined) so im not sure irr_glsl_AnisotropicViewSurfaceInteraction and irr_glsl_IsotropicViewSurfaceInteraction is the right solution

Well yeah, because `irr_glsl_IsotropicViewSurfaceInteraction` would indicate that the user does not know the Tangent vector (has not provided one) and we can make one up using the Frisvad function, and `irr_glsl_AnisotropicViewSurfaceInteraction` represents a known tangent frame where user specified `T` (and even `B`) explicitly (you cannot rotate it for them).

Reply:
I'm not against requiring bsdf sampling functions take `irr_glsl_AnisotropicViewSurfaceInteraction` as input always, I just want 3 or more different ways of turning a `irr_glsl_IsotropicViewSurfaceInteraction` into the anisotropic struct.